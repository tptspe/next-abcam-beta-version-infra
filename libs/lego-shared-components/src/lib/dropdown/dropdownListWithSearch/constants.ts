export const FILTER_MAX_RESULTS = 6
