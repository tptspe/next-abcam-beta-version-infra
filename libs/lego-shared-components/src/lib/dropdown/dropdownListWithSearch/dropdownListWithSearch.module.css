.dropdownListWithSearchContainer {
  @apply bg-white;
  @apply rounded-sm;
  @apply shadow-elevation-5;
  @apply pb-4;

  .title {
    @apply text-lg;
    @apply text-grey20;
    @apply font-sb;
    @apply p-4;
    @apply pb-2;
    @apply tracking-normal;
  }

  .searchBar {
    @apply border border-solid border-grey60;
    @apply text-grey20;
    @apply bg-interactiveGreyTransparentHover;
    @apply h-12;
    @apply mt-0 my-3 mx-4;
    @apply pl-2.5;

    width: 20.9375rem;

    &::placeholder {
      @apply text-grey60;
    }

    &:active {
      @apply border border-solid border-green45;
    }

    &:focus {
      @apply outline-none;
      @apply shadow-interactiveElement-small;
    }
  }

  .listContainer {
    .list {
      @apply flex;
      @apply flex-col;

      .listItem {
        @apply text-grey20;

        .listItemText {
          @apply py-3 px-4;

          &:focus {
            @apply outline-none;
            @apply shadow-interactiveElement-small;
          }

          &:hover {
            @apply bg-grey95;
          }
        }
      }
    }

    .noResult {
      @apply block;
      @apply mx-4 py-3;
      @apply text-grey20;
    }
  }
}
