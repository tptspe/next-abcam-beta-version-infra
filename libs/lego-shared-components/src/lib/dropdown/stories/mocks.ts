export const mockDropdownData = [
  { key: 'uk', displayValue: 'United Kingdom' },
  {
    key: 'spain',
    displayValue: 'Spain',
  },
  {
    key: 'belgium',
    displayValue: 'Belgium',
  },
  {
    key: 'netherlands',
    displayValue: 'Netherlands',
  },
  {
    key: 'france',
    displayValue: 'France',
  },
  {
    key: 'germany',
    displayValue: 'Germany',
  },
  {
    key: 'usa',
    displayValue: 'United States of America',
  },
  {
    key: 'israel',
    displayValue: 'Israel',
  },
]

export const mockDropdownDataWithShortDisplayValue = [
  { key: 'uk', displayValue: 'ğŸ‡¬ğŸ‡§ United Kingdom', shortDisplayValue: 'ğŸ‡¬ğŸ‡§' },
  {
    key: 'spain',
    displayValue: 'ğŸ‡ªğŸ‡¸ Spain',
    shortDisplayValue: 'ğŸ‡ªğŸ‡¸',
  },
  {
    key: 'belgium',
    displayValue: 'ğŸ‡§ğŸ‡ª Belgium',
    shortDisplayValue: 'ğŸ‡§ğŸ‡ª',
  },
  {
    key: 'netherlands',
    displayValue: 'ğŸ‡³ğŸ‡± Netherlands',
    shortDisplayValue: 'ğŸ‡³ğŸ‡±',
  },
  {
    key: 'france',
    displayValue: 'ğŸ‡«ğŸ‡· France',
    shortDisplayValue: 'ğŸ‡«ğŸ‡·',
  },
  {
    key: 'germany',
    displayValue: 'ğŸ‡©ğŸ‡ª Germany',
    shortDisplayValue: 'ğŸ‡©ğŸ‡ª',
  },
  {
    key: 'usa',
    displayValue: 'ğŸ‡ºğŸ‡¸ United States of America',
    shortDisplayValue: 'ğŸ‡ºğŸ‡¸',
  },
  {
    key: 'israel',
    displayValue: 'ğŸ‡®ğŸ‡± Israel',
    shortDisplayValue: 'ğŸ‡®ğŸ‡±',
  },
]
