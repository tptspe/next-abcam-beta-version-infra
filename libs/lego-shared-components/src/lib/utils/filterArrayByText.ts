type listObject = Record<string, string>
type listElement = string | listObject

export const filterArrayByText = <T extends listElement>(
  list: T[],
  keyword: string,
  filterKey?: string
): T[] => {
  const lowercaseKeyword = keyword.toLowerCase()

  return list.filter((listElem) => {
    if (listElem && filterKey && typeof listElem !== 'object') {
      throw Error(
        `You specified a filter key but didn't pass an array of object`
      )
    }

    if (listElem && !filterKey && typeof listElem === 'object') {
      throw Error(
        `You passed an array of objects but did not specify a filterKey`
      )
    }

    if (filterKey && (listElem as listObject)[filterKey]) {
      return (listElem as listObject)[filterKey]
        .toLowerCase()
        .includes(lowercaseKeyword)
    }
    return (listElem as string).toLowerCase().includes(lowercaseKeyword)
  })
}
