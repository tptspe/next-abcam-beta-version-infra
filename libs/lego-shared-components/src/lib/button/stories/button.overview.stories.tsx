import React from 'react'
import { Meta } from '@storybook/react'
import { Button } from '../button'
import { STORYBOOK_CONSTANTS } from '@abcam-web/lego-shared-components/lib/button/stories/constants'

import { Picture } from '@abcam-web/lego-shared-components/icons'

export default {
  title: 'Components/Button/Overview',
  component: Button,
} as Meta

const renderVariantGrid = (variant: any, as: any = 'button') => {
  const variantDisplayName = variant[0].toUpperCase() + variant.substring(1)
  return (
    <div className={'text-green-500'}>
      <h6>{variantDisplayName}</h6>
      <div className={'py-1 grid grid-cols-3 gap-y-4'}>
        {/*Regular sizes*/}
        {STORYBOOK_CONSTANTS.sizes.map((size: any) => {
          return (
            <Button as={as} key={size} variant={variant} size={size}>
              Button label
            </Button>
          )
        })}

        {/*Disabled*/}
        {STORYBOOK_CONSTANTS.sizes.map((size: any) => {
          return (
            <Button as={as} key={size} variant={variant} size={size} disabled>
              Button label
            </Button>
          )
        })}

        {/*iconLeft*/}
        {STORYBOOK_CONSTANTS.sizes.map((size: any) => {
          return (
            <Button
              as={as}
              variant={variant}
              size={size}
              iconLeft={<Picture />}
            >
              Button label
            </Button>
          )
        })}

        {/*iconRight*/}
        {STORYBOOK_CONSTANTS.sizes.map((size: any) => {
          return (
            <Button
              as={as}
              variant={variant}
              size={size}
              iconRight={<Picture />}
            >
              Button label
            </Button>
          )
        })}

        {/*icon both sides*/}
        {STORYBOOK_CONSTANTS.sizes.map((size: any) => {
          return (
            <Button
              as={as}
              variant={variant}
              size={size}
              iconLeft={<Picture />}
              iconRight={<Picture />}
            >
              Button label
            </Button>
          )
        })}

        {/*iconButton*/}
        {STORYBOOK_CONSTANTS.sizes.map((size: any) => {
          return (
            <Button as={as} variant={variant} size={size} iconButton>
              <Picture />
            </Button>
          )
        })}

        {/*iconButton disabled*/}
        {STORYBOOK_CONSTANTS.sizes.map((size: any) => {
          return (
            <Button as={as} variant={variant} size={size} iconButton disabled>
              <Picture />
            </Button>
          )
        })}
      </div>
    </div>
  )
}

export const asButton = () => {
  return (
    <div className={'flex flex-col items-center'}>
      {renderVariantGrid('primary')}
      {renderVariantGrid('secondary')}
      {renderVariantGrid('tertiary')}
      {renderVariantGrid('quaternary')}
      {renderVariantGrid('quinary')}
    </div>
  )
}

export const asLink = () => {
  return (
    <div className={'flex flex-col items-center'}>
      {renderVariantGrid('primary', 'a')}
      {renderVariantGrid('secondary', 'a')}
      {renderVariantGrid('tertiary', 'a')}
      {renderVariantGrid('quaternary', 'a')}
      {renderVariantGrid('quinary', 'a')}
    </div>
  )
}
