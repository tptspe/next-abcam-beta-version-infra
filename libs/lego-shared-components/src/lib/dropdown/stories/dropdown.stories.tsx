import React from 'react'
import { Meta } from '@storybook/react'
import { Dropdown } from '@abcam-web/lego-shared-components/lib'
import { DropdownListWithSearch } from '@abcam-web/lego-shared-components/lib/dropdown/dropdownListWithSearch/dropdownListWithSearch'
import { DropdownVariant } from '@abcam-web/lego-shared-components/lib/dropdown/dropdown.type'
import {
  mockDropdownData,
  mockDropdownDataWithShortDisplayValue,
} from '@abcam-web/lego-shared-components/lib/dropdown/stories/mocks'

export default {
  title: 'Components/Dropdown',
  component: Dropdown,
  argTypes: { onChange: { action: 'Value selected' } },
} as Meta

const withSearchProps = {
  title: 'Country/Region selector',
  searchPlaceholder: 'Select country/region',
} as any

const Template = (args: any) => <Dropdown {...args} />

export const WithShortDisplayValue = Template.bind({}) as any
WithShortDisplayValue.args = {
  options: mockDropdownDataWithShortDisplayValue,
  children: <DropdownListWithSearch {...withSearchProps} />,
}
WithShortDisplayValue.argTypes = {
  variant: {
    options: [DropdownVariant.primary, DropdownVariant.secondary],
    control: {
      type: 'radio',
    },
  },
}

export const WithTextValues = Template.bind({}) as any
WithTextValues.args = {
  options: mockDropdownData,
  children: <DropdownListWithSearch {...withSearchProps} />,
}
WithTextValues.argTypes = {
  variant: {
    options: [DropdownVariant.primary, DropdownVariant.secondary],
    control: {
      type: 'radio',
    },
  },
}
