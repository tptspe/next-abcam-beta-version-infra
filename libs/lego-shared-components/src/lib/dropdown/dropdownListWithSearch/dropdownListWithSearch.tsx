import React, {
  FC,
  LegacyRef,
  useCallback,
  useEffect,
  useMemo,
  useRef,
  useState,
} from 'react'
import styles from './dropdownListWithSearch.module.css'
import { DropdownOption } from '@abcam-web/lego-shared-components/lib/dropdown/dropdown.type'
import { filterArrayByText } from '@abcam-web/lego-shared-components/lib/utils/filterArrayByText'
import {
  isEnterEvent,
  keyboardVerticalNavigationHandler,
} from '@abcam-web/lego-shared-components/lib/utils/keyboardEvents'
import {
  isBlurEvent,
  isFocusEvent,
} from '@abcam-web/lego-shared-components/lib/utils/syntheticEvents'
import { FILTER_MAX_RESULTS } from '@abcam-web/lego-shared-components/lib/dropdown/dropdownListWithSearch/constants'

export type DropdownListWithSearchProps = {
  title: string
  searchPlaceholder: string
  options?: DropdownOption[]
  onValueSelect?: (value: DropdownOption) => void
}

const DropdownListWithSearch: FC<DropdownListWithSearchProps> = React.memo(
  (props) => {
    const {
      options = [],
      title,
      searchPlaceholder,
      onValueSelect = () => null,
    } = props

    const [keywordFilter, setKeywordFilter] = useState('')
    const inputFieldRef = useRef<HTMLInputElement>()

    const filteredOptions = useMemo(() => {
      return filterArrayByText(options, keywordFilter, 'displayValue')?.slice(
        0,
        FILTER_MAX_RESULTS
      )
    }, [options, keywordFilter])

    const onInputChangeCallback = useCallback(
      (evt) => {
        setKeywordFilter(evt?.target?.value)
      },
      [setKeywordFilter]
    )

    const shouldShowOptions = keywordFilter && filteredOptions
    const shouldShowNoResult = keywordFilter && filteredOptions.length === 0

    // Set the focus on the input field when the component is mounted
    useEffect(() => inputFieldRef?.current?.focus(), [])

    return (
      <div className={styles.dropdownListWithSearchContainer}>
        <h1 className={styles.title}>{title}</h1>
        <input
          className={styles.searchBar}
          ref={inputFieldRef as LegacyRef<HTMLInputElement>}
          type={'text'}
          placeholder={searchPlaceholder}
          onChange={onInputChangeCallback}
          aria-label="filter-list"
        />
        <div className={styles.listContainer}>
          {shouldShowOptions && (
            <ListRenderer
              options={filteredOptions}
              clickCallback={onValueSelect}
            />
          )}
          {shouldShowNoResult && (
            <span className={styles.noResult}>No results</span>
          )}
        </div>
      </div>
    )
  }
)

const ListRenderer = React.memo(
  (props: {
    options: DropdownOption[]
    clickCallback: (value: DropdownOption) => void
  }) => {
    const { options, clickCallback } = props
    const [selectedItemIndex, setSelectedItemIndex] = useState(0)
    const [isFocused, setIsFocused] = useState(false)

    const keyupHandler = useCallback(
      (evt) => {
        return keyboardVerticalNavigationHandler(
          evt,
          selectedItemIndex,
          setSelectedItemIndex,
          options.length - 1
        )
      },
      [selectedItemIndex, setSelectedItemIndex, options]
    )

    const focusCallback = useCallback((evt) => {
      if (isFocusEvent(evt)) {
        return setIsFocused(true)
      }
      if (isBlurEvent(evt)) {
        return setIsFocused(false)
      }
    }, [])

    return (
      <ul
        role="presentation"
        className={styles.list}
        onKeyUp={keyupHandler}
        onFocus={focusCallback}
        onBlur={focusCallback}
      >
        {options.map((option, idx) => {
          return (
            <ListItemRenderer
              key={option.key}
              autofocus={selectedItemIndex === idx && isFocused}
              option={option}
              clickCallback={clickCallback}
            />
          )
        })}
      </ul>
    )
  }
)

const ListItemRenderer = React.memo(
  (props: {
    option: DropdownOption
    clickCallback: (value: DropdownOption) => void
    autofocus: boolean
  }) => {
    const { option, clickCallback, autofocus } = props
    const itemRef = useRef<HTMLDivElement>()

    const clickHandler = useCallback(() => {
      clickCallback(option)
    }, [clickCallback, option])

    useEffect(() => {
      if (!autofocus) {
        return
      }
      itemRef?.current?.focus()
    }, [autofocus])

    const keyupHandler = useCallback(
      (evt) => {
        if (isEnterEvent(evt)) {
          clickCallback(option)
        }
      },
      [clickCallback, option]
    )

    return (
      <li
        aria-label={option.displayValue}
        key={option.key}
        className={styles.listItem}
      >
        <div
          ref={itemRef as LegacyRef<HTMLDivElement>}
          tabIndex={0}
          className={styles.listItemText}
          role={'button'}
          onClick={clickHandler}
          onKeyUp={keyupHandler}
        >
          {option.displayValue}
        </div>
      </li>
    )
  }
)

export { DropdownListWithSearch }
