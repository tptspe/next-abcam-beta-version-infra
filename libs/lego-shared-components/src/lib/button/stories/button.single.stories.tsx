import React from 'react'
import { Meta } from '@storybook/react'
import { Button } from '../button'
import { STORYBOOK_CONSTANTS } from '@abcam-web/lego-shared-components/lib/button/stories/constants'

export default {
  title: 'Components/Button/Single',
  component: Button,
  argTypes: { onClick: { action: 'clicked' } },
} as Meta

const Template = (args: any) => <Button {...args} />

export const AsButton = Template.bind({}) as any
AsButton.args = {
  variant: 'primary',
  as: 'button',
  children: 'Text',
  disabled: false,
  size: 'small',
}
AsButton.argTypes = {
  variant: {
    options: STORYBOOK_CONSTANTS.variants,
    control: {
      type: 'radio',
    },
  },
  disabled: {
    control: {
      type: 'boolean',
    },
  },
  size: {
    options: STORYBOOK_CONSTANTS.sizes,
    control: {
      type: 'radio',
    },
  },
  href: {
    table: {
      disable: true,
    },
  },
  openInNewWindow: {
    table: {
      disable: true,
    },
  },
  as: {
    table: {
      disable: true,
    },
  },
}

export const AsLink = Template.bind({}) as any
AsLink.args = {
  variant: 'primary',
  children: 'Text',
  href: 'google.com',
  openInNewWindow: false,
  size: 'small',
  as: 'a',
}
AsLink.argTypes = {
  variant: {
    control: {
      type: 'radio',
      options: STORYBOOK_CONSTANTS.variants,
    },
  },
  openInNewWindow: {
    control: {
      type: 'boolean',
    },
  },
  size: {
    control: {
      type: 'radio',
      options: STORYBOOK_CONSTANTS.sizes,
    },
  },
  as: {
    table: {
      disable: true,
    },
  },
}
