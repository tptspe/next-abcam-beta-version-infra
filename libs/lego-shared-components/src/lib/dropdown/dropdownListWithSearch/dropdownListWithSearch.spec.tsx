import React from 'react'
import {
  findByText,
  fireEvent,
  render,
  screen,
  waitFor,
} from '@testing-library/react'
import { mockDropdownData } from '@abcam-web/lego-shared-components/lib/dropdown/stories/mocks'
import { DropdownListWithSearch } from '@abcam-web/lego-shared-components/lib/dropdown/dropdownListWithSearch/dropdownListWithSearch'
import userEvent, { specialChars } from '@testing-library/user-event'
import { filterArrayByText } from '@abcam-web/lego-shared-components/lib/utils/filterArrayByText'
import { keyboardKey } from '@testing-library/user-event/dist/keyboard'

describe('The dropdownListWithSearch element', () => {
  describe('Snapshots', () => {
    it('Matches the snapshot', () => {
      const { container } = render(
        <DropdownListWithSearch
          title={'my title'}
          searchPlaceholder={'my placeholder'}
        />
      )
      expect(container).toMatchSnapshot()
    })
  })

  describe('When the user types in the search input', () => {
    it('Renders a list of options filtered by the search input value', () => {
      render(
        <DropdownListWithSearch
          title={'my title'}
          searchPlaceholder={'my placeholder'}
          options={mockDropdownData}
          onValueSelect={jest.fn()}
        />
      )
      const inputElement = screen.getByRole('textbox')
      fireEvent.change(inputElement, { target: { value: 'belg' } })
      const itemList = screen.getAllByRole('listitem')
      expect(itemList).toHaveLength(1)
    })

    it('Limits the number of results to 6', () => {
      const fakeData = []
      for (let i = 0; i < 7; i++) {
        fakeData.push({
          key: i.toString(),
          displayValue: 'eee',
        })
      }
      render(
        <DropdownListWithSearch
          title={'my title'}
          searchPlaceholder={'my placeholder'}
          options={fakeData}
          onValueSelect={jest.fn()}
        />
      )
      const inputElement = screen.getByRole('textbox')
      fireEvent.change(inputElement, { target: { value: 'e' } })
      const itemList = screen.getAllByRole('listitem')
      expect(itemList).toHaveLength(6)
    })

    it('Displays a message if there is no result', () => {
      const fakeData = []
      for (let i = 0; i < 7; i++) {
        fakeData.push({
          key: i.toString(),
          displayValue: 'eee',
        })
      }
      render(
        <DropdownListWithSearch
          title={'my title'}
          searchPlaceholder={'my placeholder'}
          options={fakeData}
          onValueSelect={jest.fn()}
        />
      )
      const inputElement = screen.getByRole('textbox')
      fireEvent.change(inputElement, { target: { value: 'a' } })
      const itemList = screen.queryAllByRole('listitem')
      expect(itemList).toHaveLength(0)
      const errorMessage = screen.findByText('No results')
      expect(errorMessage).toBeDefined()
    })
  })

  describe('When the user selects an option', () => {
    it('calls the parent callback on click', () => {
      const callback = jest.fn()
      render(
        <DropdownListWithSearch
          title={'my title'}
          searchPlaceholder={'my placeholder'}
          options={mockDropdownData}
          onValueSelect={callback}
        />
      )
      const inputElement = screen.getByRole('textbox')
      fireEvent.change(inputElement, { target: { value: 'belg' } })

      const item = screen.getByText('Belgium')
      userEvent.click(item)
      expect(callback).toHaveBeenCalled()
    })

    it('calls the parent callback on enter', () => {
      const callback = jest.fn()
      render(
        <DropdownListWithSearch
          title={'my title'}
          searchPlaceholder={'my placeholder'}
          options={mockDropdownData}
          onValueSelect={callback}
        />
      )
      const inputElement = screen.getByRole('textbox')
      fireEvent.change(inputElement, { target: { value: 'belg' } })

      const item = screen.getByText('Belgium')
      fireEvent.keyUp(item, {
        code: 'Enter',
        key: 'Enter',
      })
      expect(callback).toHaveBeenCalled()
    })
  })

  describe('Accessibility', () => {
    it('Focusses on the list when the user presses tab', () => {
      render(
        <DropdownListWithSearch
          title={'my title'}
          searchPlaceholder={'my placeholder'}
          options={mockDropdownData}
          onValueSelect={jest.fn()}
        />
      )
      const inputElement = screen.getByRole('textbox')
      fireEvent.change(inputElement, { target: { value: 'e' } })

      const firstElement = screen.getByText(mockDropdownData[0].displayValue)
      spyOn(firstElement, 'focus')

      userEvent.tab()
      expect(firstElement.focus).toHaveBeenCalled()
    })

    it('is navigable with the arrows', () => {
      render(
        <DropdownListWithSearch
          title={'my title'}
          searchPlaceholder={'my placeholder'}
          options={mockDropdownData}
          onValueSelect={jest.fn()}
        />
      )
      const inputElement = screen.getByRole('textbox')
      fireEvent.change(inputElement, { target: { value: 'e' } })

      const expectedList = filterArrayByText(
        mockDropdownData,
        'e',
        'displayValue'
      )

      const secondElement = screen.getByText(expectedList[1].displayValue)
      spyOn(secondElement, 'focus')

      userEvent.tab()
      userEvent.keyboard(`${specialChars.arrowDown}`)
      userEvent.keyboard(`${specialChars.arrowDown}`)

      expect(secondElement.focus).toHaveBeenCalled()
    })
  })
})
