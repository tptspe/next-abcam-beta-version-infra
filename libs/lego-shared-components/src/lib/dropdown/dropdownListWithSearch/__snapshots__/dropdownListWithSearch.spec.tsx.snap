// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`The dropdownListWithSearch element Snapshots Matches the snapshot 1`] = `
<div>
  <div
    class="dropdownListWithSearchContainer"
  >
    <h1
      class="title"
    >
      my title
    </h1>
    <input
      aria-label="filter-list"
      class="searchBar"
      placeholder="my placeholder"
      type="text"
    />
    <div
      class="listContainer"
    >
      
      
    </div>
  </div>
</div>
`;
